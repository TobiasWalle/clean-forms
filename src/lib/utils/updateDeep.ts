import { assertPropertyInObject } from './assertPropertyInObject';

export function updateDeep<T>(object: T, path: string[], value: any): T {
  if (path.length <= 0) {
    throw new Error('The path cannot be empty');
  }
  const lastIndex = path.length - 1;
  const tail = path.slice(0, lastIndex);
  const keyToUpdate = path[lastIndex];

  const { copy: result, selectedObject: objectToUpdate } = selectDeepAndCopy(object, tail);
  assertPropertyInObject(objectToUpdate, keyToUpdate);
  objectToUpdate[keyToUpdate] = value;

  return result;
}

function selectDeepAndCopy<T>(object: T, path: string[]): { copy: T, selectedObject: any } {
  path = path.slice(); // Copy the path
  const copy = copyArrayOrObject(object);

  let selectedObject: any = copy;
  while (path.length > 0) {
    const key = path.splice(0, 1)[0];
    assertPropertyInObject(selectedObject, key);

    const lastObject = selectedObject;
    selectedObject = copyArrayOrObject(selectedObject[key]);
    lastObject[key] = selectedObject;
  }

  return { copy, selectedObject };
}

function copyArrayOrObject<T>(object: T): T {
  if (object instanceof Array) {
    return object.slice() as any;
  } else {
    return Object.assign({}, object);
  }
}
