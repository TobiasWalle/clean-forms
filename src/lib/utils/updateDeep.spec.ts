import { updateDeep } from './updateDeep';

describe('updateDeep', () => {
  it('should deep update an object without modifying the original', () => {
    const original = { a: { b: { c: 1 } } };
    const path = ['a', 'b', 'c'];
    const value = 2;

    const newObject = updateDeep(original, path, value);

    expect(original).toEqual({ a: { b: { c: 1 } } });
    expect(newObject).toEqual({ a: { b: { c: 2 } } });
  });

  it('should deep update an object with a path of 1', () => {
    const original = { a: { b: { c: 1 } } };
    const path = ['a'];
    const value = 2;

    const newObject = updateDeep(original, path, value);

    expect(original).toEqual({ a: { b: { c: 1 } } });
    expect(newObject).toEqual({ a: 2 });
  });

  it('should throw an error if the path is 0', () => {
    const original = { a: { b: { c: 1 } } };
    const path: string[] = [];
    const value = 2;

    expect(() => updateDeep(original, path, value)).toThrowErrorMatchingSnapshot();
  });

  it('should throw an error if the path is invalid', () => {
    const original = { a: { b: { c: 1 } } };
    const path = ['b'];
    const value = 2;

    expect(() => updateDeep(original, path, value)).toThrowErrorMatchingSnapshot();
  });
});
