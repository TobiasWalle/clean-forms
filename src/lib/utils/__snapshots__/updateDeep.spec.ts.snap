// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateDeep should throw an error if the path is 0 1`] = `"The path cannot be empty"`;

exports[`updateDeep should throw an error if the path is invalid 1`] = `"The key \\"b\\" does not exits on item {\\"a\\":{\\"b\\":{\\"c\\":1}}}."`;
